README
                     OSLO - Open Secure LOader
        
 		         Bernhard  Kauer 
	                _kauer@tudos.org_


Introduction
############

OSLO, the Open Secure Loader, is a boot loader which uses the _skinit_
instruction for authenticated booting.

With around 1000 lines of code and a binary which is in the order
of 4k auditing, bugfixing and in the end trusting OSLO should be
possible.


Prerequisites
=============

To use OSLO a PC platform supporting _skinit_ is needed. There must be
a processor supporting the _skinit_ instruction, a v1.2 TPM [TPM] and
a chipset which provides memory protection for the secure loader.

The _skinit_ instruction is available with newer AMD64 processors
[AMD] and used to create a dynamic root of trust for measurement. It
is a replacement for the static root of trust provided by the BIOS and
subsequent secure boot loaders which are used on machines not
supporting skinit.


Functionality
=============

OSLO is started as kernel from a multi-boot compliant [MBI] loader. It
initialize the TPM, calls skinit, hashes [SHA] every module it gets
from the parent boot-loader and starts the first module as new kernel
in the multi-boot way.


Minimal
=======

Since OSLO is the root of trust of the secure system it is as minimized
as possible. Therefore it neither contain code to load images from
network nor hard disk. This can be done by another untrusted multi-boot
loader like GRUB or syslinux.


Components
##########

:asm.S:
  Initialize the processor on startup and after skinit by e.g. loading the
  stack pointer and segments.

:sha.c:
  A size optimized Sha1 [SHA] implementation which can hash up to 512 MB.
  Needs around 512 byte but is nearly 4 times slower than a speed
  optimized version. Since boot loading is not performance critical
  and OSLO should not hash large amount of data the speed/size
  tradeoff is acceptable here.

:tis.c:
  A simple TIS [TIS] driver using the memory mapped interface of
  version 1.2 TPMs. Tested with TPMs from Infineon, STM, Atmel and
  Broadcom.

:tpm.c:
  The needed TPM functions, like TPM_Extend.

:elf.c:
  The elf decoding.

:osl.c:
  The main program including hashing the modules and
  startup of the first one.  

:util.c asm.h:
  Helper functions for string output and low level hardware access
  like _rdmsr_.

:mp.c mp.h:
  Helper functions to start and stop processors on an MP system.


:beirut.c:
  A helper program that hashes the command line of other multiboot
  modules.

:pamplona.c:
  A helper program that does everything to reverse the steps done by
  OSLO. For example it removes DEV protection and clears the global
  interrupt flag. It does allow you to use OSLO but start an
  unmodified OS in an unsecure way after that.


:munich.c:
  A helper program to start Linux [LBP] from a multiboot compliant
  loader. The first module is used as linux kernel. The optional
  second one is used as initrd.


FAQ
###

: Why is DEV protection not implemented? :
  Mainly time reasons and the realization that the workaround for the
  DEV init race has to be implemented.

: Why do you extend the PCR for every module? :
  Previously OSLO would calc a hash over all multiboot modules. This
  was a bug since hashing one module or hashing two halves of the same
  binary with two modules was indistinguishable. The hole between the
  two halves could be used by an attacker to hide untrusted code.

: Why do you implemented Beirut? :
  There are applications which can be fooled by giving them a faked
  commandline. Since someone decided not to fix the applications
  and handle untrusted command lines like untrusted config files
  (e.g. simple hash them itself), we have to hash the commandline
  before we pass it to them. Or in other words: Beirut is just a nasty
  hack for compatibility reasons.

: Is OSLO bugfree? :
  Of course not in this version. Currently at least two bugs are
  known: The DEV initialization is missing and the famous "type of
  memory" bug is still present.


References
##########

:AMD:  AMD64 Architecture Programmer's Manual Vol. 2 - http://www.amd.com/us-en/assets/content_type/white_papers_and_tech_docs/24593.pdf 
:LBP:  Linux/I386 Boot Protocol - http://lxr.linux.no/source/Documentation/i386/boot.txt
:MBI:  multiboot specification  - http://www.gnu.org/software/grub/manual/multiboot/multiboot.txt
:SHA:  US Secure Hash Algorithm 1 (SHA1) - http://www.ietf.org/rfc/rfc3174.txt
:TIS:  TPM Interface Specification - https://www.trustedcomputinggroup.org/groups/pc_client/TCG_PCClientTPMSpecification_1-20_1-00_FINAL.pdf
:TPM:  TCG TPM Specification v1.2 rev 5 Part 3 Commands - https://www.trustedcomputinggroup.org/groups/tpm/mainP3Commands_rev85.zip 


Acknowledgements
################

OSLO was funded by OpenTC. Special thanks go to AMD Dresden for the
constant support and Infineon Munich for the help with TPMs.


CHANGELOG
OSLO (0.4.5)

* fix tis_access return code check in Beirut to work with Atmel TPMs
* fix reboot function

 -- Bernhard Kauer <kauer@os.inf.tu-dresden.de>  Di, 19 June 2007


OSLO (0.4.4)

* make tis_access() portable - thanks to Broadcom
* add support for Qemu TPM
* made it more usable for the linux wrapper tis_oslo

 -- Bernhard Kauer <kauer@os.inf.tu-dresden.de>  Di, 13 March 2007


OSLO (0.4.3)

  * support for Broadcom TPMs - thanks to Jonathan M. McCune for debugging
  * use another region for realmode code to not clash with PXE loader
  * improved 64bit cleanness

 -- Bernhard Kauer <kauer@os.inf.tu-dresden.de>  Mo, 26 February 2007


OSLO (0.4.2)

  * fix initialization of serial devices
  * add pamplona code to start unmodified OSes
  * fix multiboot header

 -- Bernhard Kauer <kauer@os.inf.tu-dresden.de>  Mo, 13 November 2006


OSLO (0.4.1)

  * fix nasty SHA1 bug, triggered on hashing cmdlines
  * add fixup code for buggy Atmel TPMs
  * add serial output for debugging

 -- Bernhard Kauer <kauer@os.inf.tu-dresden.de>  Fr, 20 October 2006


OSLO (0.4.0)

  * execute MP fixup code on multiprocessor platforms
  * fixes the "ACPI enable" bug

 -- Bernhard Kauer <kauer@os.inf.tu-dresden.de>  Fr, 14 July 2006


OSLO (0.3.5)

  * added MUNICH to start Linux from a multiboot loader

 -- Bernhard Kauer <kauer@os.inf.tu-dresden.de>  Wed, 05 July 2006


OSLO (0.3.4)

  * added BEIRUT to calc hashes of command lines
  * output a message label ("OSLO:") to distuingish different apps
  * move start_module into elf.c to allow sharing with BEIRUT

 -- Bernhard Kauer <kauer@os.inf.tu-dresden.de>  Wed, 07 June 2006


OSLO (0.3.3)

  * do not give the locality with every command
  * make header files ANSI-C compatible
  * set bootloader name in mbi structure
  * bugfix: calc a hash for every module

 -- Bernhard Kauer <kauer@os.inf.tu-dresden.de>  Tue, 30 May 2006

OSLO (0.3.2)

  * poll the PIT to wait a number of milliseconds

 -- Bernhard Kauer <kauer@os.inf.tu-dresden.de>  Thu, 27 Apr 2006


OSLO (0.3.1)

  * implement memcpy and memset to let it compile with gcc-3.4
  * use -ffunction-sections and -gc-section to remove unused code

 -- Bernhard Kauer <kauer@os.inf.tu-dresden.de>  Wed, 19 Apr 2006


OSLO (0.3)

  * support for multiprocessor platforms
  * fix "SL length multiple of 4" bug
  * renamed to OSLO because of naming conflicts

 -- Bernhard Kauer <kauer@os.inf.tu-dresden.de>  Fri, 7 Apr 2006


OSL (0.2)

  * initial public release

 -- Bernhard Kauer <kauer@os.inf.tu-dresden.de>  Thu, 28 Mar 2006


osl.c
/*
 * \brief   OSLO - Open Secure Loader
 * \date    2006-03-28
 * \author  Bernhard Kauer <kauer@tudos.org>
 */
/*
 * Copyright (C) 2006-2007  Bernhard Kauer <kauer@tudos.org>
 * Technische Universitaet Dresden, Operating Systems Research Group
 *
 * This file is part of the OSLO package, which is distributed under
 * the  terms  of the  GNU General Public Licence 2.  Please see the
 * COPYING file for details.
 */


#include "version.h"
#include "util.h"
#include "sha.h"
#include "elf.h"
#include "tpm.h"
#include "mp.h"
#include "osl.h"

const char *version_string = "OSLO " VERSION "\n";
const char *message_label = "OSLO:   ";

/**
 * Function to output a hash.
 */
static void
show_hash(char *s, unsigned char *hash)
{
  out_string(message_label);
  out_string(s);
  for (unsigned i=0; i<20; i++)
    out_hex(hash[i], 7);
  out_char('\n');
}


/**
 *  Hash all multiboot modules.
 */
static
int
mbi_calc_hash(struct mbi *mbi, struct Context *ctx)
{
  unsigned res;

  CHECK3(-11, ~mbi->flags & MBI_FLAG_MODS, "module flag missing");
  CHECK3(-12, !mbi->mods_count, "no module to hash");
  out_description("Hashing modules count:", mbi->mods_count);

  struct module *m  = (struct module *) (mbi->mods_addr);
  for (unsigned i=0; i < mbi->mods_count; i++, m++)
    {
      sha1_init(ctx);
      CHECK3(-13, m->mod_end < m->mod_start, "mod_end less than start");
      sha1(ctx, (unsigned char*) m->mod_start, m->mod_end - m->mod_start);
      sha1_finish(ctx);
      CHECK4(-14, (res = TPM_Extend(ctx->buffer, 19, ctx->hash)), "TPM extend failed", res);
    }
  return 0;
}


/**
 * Prepare the TPM for skinit.
 * Returns a TIS_INIT_* value.
 */
static
int
prepare_tpm(unsigned char *buffer)
{
  int tpm, res;

  CHECK4(-60, 0 >= (tpm = tis_init(TIS_BASE)), "tis init failed", tpm);
  CHECK3(-61, !tis_access(TIS_LOCALITY_0, 0), "could not gain TIS ownership");
  if ((res=TPM_Startup_Clear(buffer)) && res!=0x26)
    out_description("TPM_Startup() failed", res);

  CHECK3(-62, tis_deactivate_all(), "tis_deactivate failed");
  return tpm;
}


/**
 * This function runs before skinit and has to enable SVM in the processor
 * and disable all localities.
 */
int
__main(struct mbi *mbi, unsigned flags)
{

  unsigned char buffer[TCG_BUFFER_SIZE];

#ifndef NDEBUG
  serial_init();
#endif
  out_string(version_string);
  ERROR(10, !mbi || flags != MBI_MAGIC, "not loaded via multiboot");

  // set bootloader name
  mbi->flags |= MBI_FLAG_BOOT_LOADER_NAME;
  mbi->boot_loader_name = (unsigned) version_string;

  int revision = 0;
  if (0 >= prepare_tpm(buffer) || (0 > (revision = check_cpuid())))
    {
      if (0 > revision)
	out_info("No SVM platform");
      else
	out_info("Could not prepare the TPM");

      ERROR(11, start_module(mbi), "start module failed");
    }

  out_description("SVM revision:", revision);
  ERROR(12, enable_svm(), "could not enable SVM");

  /**
   * All APs have to be in the INIT state before skinit can be
   * executed.
   */
  ERROR(13, stop_processors(), "sending an INIT IPI to other processors failed");

  wait(1000);
  out_info("call skinit");
  do_skinit();
}


/**
 * This code is executed after skinit.
 */
int
oslo(struct mbi *mbi)
{
  struct Context ctx;
  
  ERROR(20, !mbi, "no mbi in oslo()");

  if (tis_init(TIS_BASE))
    {
      ERROR(21, !tis_access(TIS_LOCALITY_2, 0), "could not gain TIS ownership");
      ERROR(22, mbi_calc_hash(mbi, &ctx),  "calc hash failed");
      show_hash("PCR[19]: ",ctx.hash);

#ifndef NDEBUG
      int res;
      dump_pcrs(ctx.buffer);

      CHECK4(24,(res = TPM_PcrRead(ctx.buffer, 17, ctx.hash)), "TPM_PcrRead failed", res);
      show_hash("PCR[17]: ",ctx.hash);
#endif
      ERROR(25, tis_deactivate_all(), "tis_deactivate failed");
  }
  ERROR(27, start_module(mbi), "start module failed");
  return 28;
}
